src/core/net/dnssd_server.cpp

```diff
@@ -0,0 +1,646 @@
+
+    if (nameOffset != NameCompressInfo::kUnknownOffset)
+    {
+        error = Dns::Name::AppendPointerLabel(nameOffset, aMessage);
+    }
+    else
+    {
+        uint8_t serviceStart = static_cast<uint8_t>(StringLength(aName, Name::kMaxLength) -
+                                                    StringLength(aCompressInfo.GetServiceName(), Name::kMaxLength));
+
+        aCompressInfo.SetInstanceNameOffset(aMessage.GetLength(), aName);
+
+        SuccessOrExit(error = Dns::Name::AppendLabel(aName, serviceStart - 1, aMessage));
+        error = Dns::Name::AppendPointerLabel(aCompressInfo.GetServiceNameOffset(), aMessage);
+    }
+
+exit:
+    return error;
+}
+
+otError Server::AppendHostName(Message &aMessage, const char *aName, NameCompressInfo &aCompressInfo)
+{
+    otError  error;
+    uint16_t hostCompressOffset = aCompressInfo.GetHostNameOffset(aName);
+
+    if (hostCompressOffset != NameCompressInfo::kUnknownOffset)
+    {
+        error = Dns::Name::AppendPointerLabel(hostCompressOffset, aMessage);
+    }
+    else
+    {
+        uint8_t domainStart = static_cast<uint8_t>(StringLength(aName, Name::kMaxLength) -
+                                                   StringLength(aCompressInfo.GetDomainName(), Name::kMaxLength));
+
+        aCompressInfo.SetHostNameOffset(aMessage.GetLength(), aName);
+
+        SuccessOrExit(error = Dns::Name::AppendMultipleLabels(aName, domainStart, aMessage));
+        error = Dns::Name::AppendPointerLabel(aCompressInfo.GetDomainNameOffset(), aMessage);
+    }
+
+exit:
+    return error;
+}
+
+void Server::IncResourceRecordCount(Header &aHeader, bool aAdditional)
+{
+    if (aAdditional)
+    {
+        aHeader.SetAdditionalRecordCount(aHeader.GetAdditionalRecordCount() + 1);
+    }
+    else
+    {
+        aHeader.SetAnswerCount(aHeader.GetAnswerCount() + 1);
+    }
+}
+
+otError Server::FindNameComponents(const char *aName, const char *aDomain, NameComponentsOffsetInfo &aInfo)
+{
+    uint8_t nameLen   = static_cast<uint8_t>(StringLength(aName, Name::kMaxLength));
+    uint8_t domainLen = static_cast<uint8_t>(StringLength(aDomain, Name::kMaxLength));
+    otError error     = OT_ERROR_NONE;
+    uint8_t labelStart, labelStop;
+
+    // make sure name ends with domain
+    VerifyOrExit(domainLen > 0 && nameLen >= domainLen, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(memcmp(&aName[nameLen - domainLen], aDomain, domainLen) == 0, error = OT_ERROR_INVALID_ARGS);
+
+    labelStart          = nameLen - domainLen;
+    aInfo.mDomainOffset = labelStart;
+
+    while (true)
+    {
+        error = FindPreviousLabel(aName, labelStart, labelStop);
+
+        VerifyOrExit(error == OT_ERROR_NONE, error = (error == OT_ERROR_NOT_FOUND ? OT_ERROR_NONE : error));
+
+        if (labelStop == labelStart + 4 && (memcmp(&aName[labelStart], kDnssdProtocolUdp, 4) == 0 ||
+                                            memcmp(&aName[labelStart], kDnssdProtocolTcp, 4) == 0))
```

Comment:
```suggestion
                                            memcmp(&aName[labelStart], kDnssdProtocolTcp, sizeof(kDnssdProtocolTcp)) == 0))
```

Reply:
Defined a new constant `kProtocolLabelLength = 4`. 

I intended not to use `sizeof(kDnssdProtocolTcp)` because if we changed the implementation to include '\0' in `kDnssdProtocolTcp`, the size would then be 5. I tend to keep the code more robust against potential future changes.